rust/publish-ghost/src/lib.rs

```diff
@@ -1,9 +1,122 @@
+    eyre::{bail, eyre, Context, Result}, 
+    reqwest::Client,
+    serde::{Deserialize, Serialize},
+    serde_json::{self, json},
+    tracing,
 };
+use document::{CommandWait, Document};
+
+use url::Host;
+use jsonwebtoken as jwt;
+
+const SECRET_NAME: &str = "GHOST_ADMIN_API_KEY";
+
+// use eyre::{eyre, Result};
+use std::time::{SystemTime, UNIX_EPOCH};
+
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(crate = "common::serde")]
+struct Claims {
+    // "Audience", e.g. a URL in the Ghost instance
+    aud: String,
+    iat: u64,
+    exp: u64,
+}
+
+fn generate_jwt(key: &str) -> Result<String> {
+    let Some((id, secret)) =  key.split_once(':') else {
+        return Err(eyre!("invalid Ghost Admin API key")); // should never happen
+    };
+
+    let iat = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .map_err(|e| eyre!("error accessing system clock: {}", e))?
+        .as_secs();
+
+    let exp = iat + 5 * 60; // 5 minutes
+    let aud = "/admin/".to_string();
+
+    let mut header = jwt::Header::new(jwt::Algorithm::HS256);
+    header.typ = Some("JWT".to_string());
+    header.kid = Some(id.to_string());
+    
+    let payload = Claims {
+        aud,
+        iat,
+        exp,
+    };
+
+    let secret: Result<Vec<u8>> = secret
+        .as_bytes()
+        .chunks(2)
+        .map(|chunk| {
+            let hex_pair = std::str::from_utf8(chunk)?; // will always succeed as we start with a str
+
+            u8::from_str_radix(hex_pair, 16)
+                .map_err(|e| eyre!("invalid input in secret: {}", e))
+        })
+        .collect();
+    let secret = secret?;
+
+    let secret = jwt::EncodingKey::from_secret(&secret);
+    let token = jwt::encode(&header, &payload, &secret)?;
+
+    Ok(token)
+}
+
+fn only_hex(s: &str) -> bool {
+    s.chars().all(|c| c.is_ascii_lowercase() && c.is_ascii_hexdigit())
+}
+
+// Validate that key looks like a Ghost Admin API key
+fn validate_key(key: &str) -> Result<String> {
+    // Split into id:secret
+    let Some((id, secret)) = key.split_once(':') else {
+        bail!("Ghost Admin API key must be in format `id:secret`, i.e. an id and secret separated by a colon.");
+    };
+
+    if id.is_empty() {
+        bail!("The id field of `key` must not be empty");
+    }
+    if secret.is_empty() {
+        bail!("The secret field of `key` must not be empty");
+    }
+
+    if !only_hex(id) || !only_hex(secret) {
+        tracing::warn!("Ghost Admin API key may be invalid; should only contain lowercase hexadecimal characters");
+    }
+
+    Ok(key.to_string())
+}
+
+/// Parse an input from the command line as a Ghost Admin API key
+fn parse_key(arg: &str) -> Result<String> {
+    // Use the key provided on the command-line
+    if !arg.is_empty() {
+        return validate_key(arg);
+    }
+
+    // If not, check if it's provided as an environment variable
+    if let Ok(env_key) = std::env::var("STENCILA_GHOST_KEY") {
+        return validate_key(&env_key);
+    }
```

Comment:
I thought that using `#[arg(..., env=)]` that we get this "for free" - i.e. CLAP checks for the env var. Or maybe it is not done before `parse_arg`

Reply:
I actually wasn't sure when I wrote it, so decided to include it so that the order is correct. I was worried that the env var would only be accessed by clap if `parse_arg` failed to find anything.

Keeping it in the attribute means that the documentation is rendered correctly.

Reply:
Good thinking.